### PR TITLE
Use docker-compose, consume APIs, show cases and referrals on dashboard

### PR DESCRIPTION
**Pull request by eckdanny-osi**
closes [CASE-1041](https://osi-cwds.atlassian.net/browse/CASE-1041), [CASE-1025](https://osi-cwds.atlassian.net/browse/CASE-1025)

Adds `db2data` and `case-api` images to the `docker-compose` environment. Adds route handling for cases and referrals. Refactors a table component.

### CODE REVIEW

**Comment by eckdanny-osi** [comment]
started and pushed for early feedback. Updates `compose` configs to include db2 and api images.

File: app/controllers/api/cases_controller.rb [hunk 1]
```diff
@@ -3,21 +3,16 @@
 module Api
   class CasesController < ActionController::API
     def cases_by_user
-      cases = names.map.with_index do |name, index|
-        create_case(index + 1, name)
+      conn = Faraday.new(:url => 'http://case-api:8080') do |faraday|
```
Comment (eckdanny-osi) [inline app/controllers/api/cases_controller.rb]: _Help!_ I've gone full-circle with code climate trying to resolve this... Get either "use new Hash syntax" or "unnecessary curlies"

File: app/javascript/_components/Table/Table.test.js [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+        .first()
+        .text()
+    ).toEqual('Foo');
+  });
+
+  it('renders data in the tbody', () => {
+    const myData = [['a1', 'a2', 'a3'], ['b1', 'b2', 'b3'], ['c1', 'c2', 'c3']];
+    const wrapper = shallow(<Table colNames={[]} data={myData} />);
+    expect(wrapper.find('tbody tr').length).toBe(3);
+    const tds = wrapper
```
Comment (RamuKammagani-osi) [inline app/javascript/_components/Table/Table.test.js]: I think we are not expected to use short cuts like (const tds = wrapper)! Can you check once again?

Comment (eckdanny-osi) [reply app/javascript/_components/Table/Table.test.js]: no shortcut. `<td>` pluralized

Comment (RamuKammagani-osi) [reply app/javascript/_components/Table/Table.test.js]: Then 👍

**Comment by eckdanny-osi** [comment]
The developers were tricked up a little bit because we all have gotten used to `yarn start`. I did not make it clear that starting up the dev environment should now be done with `docker-compose up`. 

Consider a follow up PR to add similar _convenience-style_ startup hook in yarn for consistency's sake.

This feels appropriate to me: `developer` <--`yarn`--> `env`. (Disclaimer: I'm primarily a JS guy so node-based CLIs are natural to me)

File: app/javascript/dashboard/DashboardContainer.jsx [hunk 3]
```diff
@@ -2,6 +2,13 @@ import React from 'react';
 import axios from 'axios';
 import { GlobalHeader, PageHeader, Cards, Alert } from 'react-wood-duck';
 import Caseload from '../_components/Caseload';
+import Table from '../_components/Table';
+import { transformCase, transformReferral } from './transforms';
+
+/**
+ * @todo: Use a service to manage the user profile when one exists
+ */
+const getUserId = () => '0Ki';
```
Comment (edgibbs) [inline app/javascript/dashboard/DashboardContainer.jsx]: what is 0ki

File: config/routes.rb [hunk 4]
```diff
@@ -4,11 +4,17 @@
...
-  namespace :api do
+  namespace :api, defaults: {format: 'json'} do
     resources :cases, only: [ :index ] do
       collection do
         get ':user_id/index', to: 'cases#cases_by_user'
       end
     end
+
+    resources :referrals, only: [ :index ] do
+      collection do
```
Comment (edgibbs) [inline config/routes.rb]: No route test exists for referrals

File: app/javascript/dashboard/DashboardContainer.jsx [hunk 5]
```diff
@@ -12,24 +19,63 @@ class DashboardContainer extends React.Component {
...
+      .then(cases => cases.map(transformCase))
+      .catch(err => {
+        throw err;
+      });
+  };
+
+  renderReferrals = () => {
+    if (!this.state.referrals) return false;
+    const data = this.state.referrals.reduce(
+      (aggr, { id, name, assignmentType }) => {
```
Comment (edgibbs) [inline app/javascript/dashboard/DashboardContainer.jsx]: Never seen `aggr`, prefer `accumulator` here as a variable name, but have seen a pretty common pattern of `acc` as well.

File: app/javascript/dashboard/DashboardContainer.test.js [hunk 6]
```diff
@@ -0,0 +1,106 @@
...
+        });
+        const wrapper = shallow(<DashboardContainer />);
+        wrapper
+          .instance()
+          .fetchReferrals()
+          .then(() => {
+            const referrals = wrapper.state('referrals');
+            expect(referrals.records.length).toEqual(3);
+          });
+      });
```
Comment (edgibbs) [inline app/javascript/dashboard/DashboardContainer.test.js]: Doesn't test the error case.

File: app/javascript/dashboard/DashboardContainer.test.js [hunk 7]
```diff
@@ -0,0 +1,106 @@
...
+        const wrapper = shallow(<DashboardContainer />);
+        wrapper
+          .instance()
+          .fetchCases()
+          .then(() => {
+            const caseload = wrapper.state('caseload');
+            expect(caseload).toBeDefined();
+            expect(caseload.XHRStatus).toEqual('ready');
+            expect(caseload.records.length).toEqual(3);
+          });
```
Comment (edgibbs) [inline app/javascript/dashboard/DashboardContainer.test.js]: Doesn't test the error case

File: .env [hunk 8]
```diff
@@ -0,0 +1,13 @@
+#DB2
```
Comment (edgibbs) [inline .env]: I think we might want to avoid checking this in.

File: app/javascript/dashboard/DashboardContainer.jsx (hunk 3 above)
Comment (eckdanny-osi) [reply app/javascript/dashboard/DashboardContainer.jsx]: `0ki` is the staff_id in the swagger dashboard requests. (It has cases and referrals in db2data container)

File: app/javascript/dashboard/DashboardContainer.jsx (hunk 5 above)
Comment (eckdanny-osi) [reply app/javascript/dashboard/DashboardContainer.jsx]: i'm down with `acc`

**Comment by eckdanny-osi** [comment]
I'm going to back out the _payload transform_ business... I'd **much** prefer to have a consistent structural types for a domain models like `case` and `referral` across the stack.

Reverting will immediately raise lint errors like

> Identifier 'active_service_component' is not in camel case.

but in the long run I feel strongly its for the best. We can either relax the linting rule _or_ determine (finally) whether camelCase payloads should be the norm.

**Comment by eckdanny-osi** [comment]
I plan to unwind the `_types` ts <-- > jsdoc in favor of *.d.ts files to provide interfaces on classes/modules/functions. A bit cleaner. Not as easy as working in actual typescript. For use cases like type-hinting for API payloads, removes some of the pain.

screen shot 2018-01-04 at 1 30 15 am

File: app/javascript/dashboard/DashboardContainer.jsx (hunk 3 above)
Comment (edgibbs) [reply app/javascript/dashboard/DashboardContainer.jsx]: ahh

File: app/javascript/_components/Caseload/Caseload.test.jsx [hunk 9]
```diff
@@ -30,10 +31,9 @@ describe('Caseload', () => {
...
     });
 
-    it('should return a table', () => {
+    it('should return a Table', () => {
       const Wrapper = shallow(<Caseload status="ready" cases={cases} />);
-      const children = Wrapper.find('tbody > tr');
-      expect(children.length).toBe(cases.length);
+      expect(Wrapper.find(Table).exists()).toBe(true);
     });
   });
```
Comment (edgibbs) [inline app/javascript/_components/Caseload/Caseload.test.jsx]: Looks like it's missing a test for the title showing Cases or Cases with (#) for the headers.

File: package.json [hunk 10]
```diff
@@ -31,8 +33,11 @@
     "sass-loader": "^6.0.6"
   },
   "devDependencies": {
+    "@types/axios": "^0.14.0",
```
Comment (eckdanny-osi) [inline package.json]: dont actually need this

File: config/environments/development.rb [hunk 11]
```diff
@@ -48,6 +48,8 @@
   # Raises error for missing translations
   # config.action_view.raise_on_missing_translations = true
 
+  config.log_level = :warn
```
Comment (eckdanny-osi) [inline config/environments/development.rb]: undo

File: package.json [hunk 12]
```diff
@@ -8,16 +8,18 @@
...
     "test:rspec": "bundle exec rspec",
     "test:jest:watch": "npm run test:jest -- --watch",
-    "lint:test:eslint": "eslint --format tap \"app/javascript/**/*.{js,jsx}\"",
-    "lint:test:rubocop": "bundle exec rubocop --format tap",
-    "lint:test": "run-p --print-label lint:test:*",
-    "lint:configtest": "eslint --print-config .eslintrc.js | eslint-config-prettier-check",
-    "lint": "npm-run-all lint:configtest lint:test",
+    "lint:eslint": "eslint --format tap \"app/javascript/**/*.{js,jsx}\"",
+    "lint:eslint:chkconfig": "eslint --print-config .eslintrc.js | eslint-config-prettier-check",
+    "lint:LKSDJFLKSDJF": "run-p --print-label lint:test:*",
```
Comment (eckdanny-osi) [inline package.json]: fix

File: .gitignore [hunk 13]
```diff
@@ -32,3 +32,6 @@ package-lock.json
 *.sublime-workspace
 *.sublime-project
 /.vscode
+
+# ENV
```
Comment (eckdanny-osi) [inline .gitignore]: remove

File: app/controllers/api/referrals_controller.rb [hunk 14]
```diff
@@ -0,0 +1,23 @@
+# frozen_string_literal: true
+
+module Api
+  class ReferralsController < ActionController::API
+    def referrals_by_user
+      referrals = Referral::ReferralRepository.new.referrals_by_user_id params[:user_id]
+      puts referrals
```
Comment (eckdanny-osi) [inline app/controllers/api/referrals_controller.rb]: remove `puts`

File: app/controllers/api/referrals_controller.rb [hunk 15]
```diff
@@ -0,0 +1,23 @@
...
+      render json: referrals
+    end
+    # def referrals_by_user
```
Comment (eckdanny-osi) [inline app/controllers/api/referrals_controller.rb]: remove

File: app/javascript/dashboard/DashboardContainer.jsx [hunk 16]
```diff
@@ -1,35 +1,128 @@
...
+          cases: {
+            XHRStatus: 'ready',
+            records: cases,
+          },
+        })
+      )
+      .catch(() => this.setState({ cases: { XHRStatus: 'error' } }));
+  };
+
+  renderCases = () => {
```
Comment (eckdanny-osi) [inline app/javascript/dashboard/DashboardContainer.jsx]: this and `renderReferrals` are so similar. Refactor.

File: config/micro_services.yml [hunk 17]
```diff
@@ -1,5 +1,5 @@
 development:
-  case_api_base_url: 'https://casemgmapi.dev.cwds.io'
+  case_api_base_url: 'http://case-api:8080' # 'https://casemgmapi.dev.cwds.io'
```
Comment (eckdanny-osi) [inline config/micro_services.yml]: @edgibbs  nope. We're not going to rely on docker networking... so should this be an ENV var?

File: lib/case/case_repository.rb [hunk 18]
```diff
@@ -7,8 +7,11 @@ def initialize(http_service = Infrastructure::HttpService.new)
     end
 
     def cases_by_user_id(user_id)
-      raw_json = @http_service.get("/cases/_search=user=#{user_id}")
-      JSON.parse(raw_json, symbolize_names: true)[:results].map { |result| Case.new(result) }
+      response = @http_service.get("/staff/#{user_id}/cases")
+      # JSON.parse(response.body, symbolize_names: true)[:results].map { |result| Case.new(result) }
+      # @(ed): y u no work???
+      # response.body.map { |result| Case.new(result) }
```
Comment (eckdanny-osi) [inline lib/case/case_repository.rb]: @edgibbs i needs ruby help marshaling json into instances.

File: lib/referral/referral_repository.rb [hunk 19]
```diff
@@ -0,0 +1,17 @@
...
+module Referral
+  class ReferralRepository
+    def initialize(http_service = Infrastructure::HttpService.new)
+      @http_service = http_service
+    end
+
+    def referrals_by_user_id(user_id)
+      response = @http_service.get("/staff/#{user_id}/referrals")
+      # JSON.parse(response.body, symbolize_names: true)[:results].map { |result| Case.new(result) }
+      # @(ed): y u no work???
```
Comment (eckdanny-osi) [inline lib/referral/referral_repository.rb]: @edgibbs i needs ruby help marshaling json into instances.

File: app/javascript/_components/Table/Table.test.js [hunk 20]
```diff
@@ -0,0 +1,38 @@
...
+      .find('tbody tr')
+      .first()
+      .find('td');
+    expect(tds.length).toBe(3);
+    expect(tds.first().text()).toEqual('a1');
+  });
+});
```
[the remark this answers is not in the document]
Comment (eckdanny-osi) [reply app/javascript/_components/Table/Table.test.js]: dealing with empty resultset is a concern for higher level component.

File: app/javascript/dashboard/DashboardContainer.jsx [hunk 21]
```diff
@@ -73,3 +114,9 @@ class DashboardContainer extends React.Component {
 }
 
 export default DashboardContainer;
+
+function getCardHeaderText({ XHRStatus, records }, text = 'Records') {
+  return XHRStatus === 'ready' && records && records.length
+    ? `${text} (${records.length})`
+    : text;
```
Comment (eckdanny-osi) [inline app/javascript/dashboard/DashboardContainer.jsx]: test both branches

File: app/javascript/dashboard/DashboardContainer.test.js [hunk 22]
```diff
@@ -0,0 +1,92 @@
...
+      const instance = wrapper.instance();
+      instance.fetchCases();
+    });
+
+    it('tracks referrals api requests', () => {
+      ReferralService.fetch.mockReturnValue(Promise.reject(Error('error')));
+      const wrapper = shallow(<DashboardContainer />);
+      const instance = wrapper.instance();
+      instance.fetchReferrals();
+      console.log(wrapper.state());
```
Comment (eckdanny-osi) [inline app/javascript/dashboard/DashboardContainer.test.js]: remove

**Comment by edgibbs** [comment]
Code climate issues are accepted as small technical debt and false positives.

**Pull request merged**